drivers: timer: nrf_grtc_timer: Optimize to reduce register access

nordic-krch opened the pull request:
Speed up execution of the interrupt handler and sys_clock_set_timeout(). Sys_clock_set_timeout() can be called in two scenarios: from previous timeout expiration handler or freely. If the former case fast path can be used since CC value in the GRTC register just expired and it can be used as a reference for CCADD setting. This is only a single register write so it's much faster. In the latter a longer procedure is applied which also happens in two variants. If value which is set in CC is further in the future (e.g. K_FOREVER was set before) then CC can be safely overwritten with a new value without a risk of triggering unexpected COMPARE event. If value in CC is earlier than the new CC value (if earlier timeout was aborted) then there is a risk of COMPARE event happening while it is being overwritten. That case requires long and safer procedure of setting CC.

Update hal_nordic with changes in the nrfx_grtc driver which are needed for nrf_grtc_timer changes.

I've run a simple test which measures CPU load during
```
for (int i = 0; i < 10; i++) {
  k_msleep(1);
}
```
And I've got following results (note that there are some kernel timeout optimizations proposed that improve even more):
| Soc  | before | after | + #87948 #87816 |
| ----- | ----------|------- |------------------------ |
| nrf54l15l  | 3.8  | 2.9 | 1.5 |
| nrf54h20  | 3.0  | 1.9 | 1.6 |

nordic-krch committed:
drivers: timer: nrf_grtc_timer: Optimize to reduce register access

Speed up execution of the interrupt handler and sys_clock_set_timeout().
Sys_clock_set_timeout() can be called in two scenarios: from previous
timeout expiration handler or freely. If the former case fast path
can be used since CC value in the GRTC register just expired and it
can be used as a reference for CCADD setting. This is only a single
register write so it's much faster. In the latter a longer procedure
is applied which also happens in two variants. If value which is
set in CC is further in the future (e.g. K_FOREVER was set before) then
CC can be safely overwritten with a new value without a risk of
triggering unexpected COMPARE event. If value in CC is earlier than
the new CC value (if earlier timeout was aborted) then there is a
risk of COMPARE event happening while it is being overwritten.
That case requires long and safer procedure of setting CC.

Update hal_nordic with changes in the nrfx_grtc driver which are
needed for nrf_grtc_timer changes.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
fyi @adamkondraciuk

nordic-krch committed:
tests: drivers: timer: nrf_grtc_timer: Add stress test

Add stress test that randomly starts and aborts multiple timers
from various contexts. Test checks if timers do not expire prematurely.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

andyross commented on drivers/timer/nrf_grtc_timer.c:
```diff
@@ -551,22 +553,47 @@ void sys_clock_set_timeout(int32_t ticks, bool idle)
+	uint32_t ch = system_clock_channel_data.channel;
 
-	uint64_t delta_time = ticks * CYC_PER_TICK;
+	if ((cc_value == expired_cc) && (ticks < MAX_REL_TICKS)) {
```
FWIW, for the specific case you're trying to handle here (re-setting the timeout at the end of the handler), you can detect that even faster by seeing that you're being called inside the timer ISR.  Set a static boolean before calling the handler and clear it after it returns.  That would spend the two extra memory operations for every ISR, but would avoid having to call cc_value_read() on every call to set_timeout, which is almost certainly a win.

nordic-krch replied:
I had that approach initially, but it was failing in a stress test which is calling calling k_timer from interrupt context which has higher priority than the system timer interrupt and I had to change it to this condition so only first attempt that follows timeout expiration is using optimize setting.

andyross reviewed (dismissed):
Looks clean.  One suggestion for further optimization/simplification.

nordic-segl commented on tests/drivers/timer/nrf_grtc_timer/testcase.yaml:
```diff
@@ -1,21 +1,24 @@
+    - nrf54h20dk/nrf54h20/cpuppr
+    - nrf54l20pdk/nrf54l20/cpuapp
+    - nrf54lm20dk/nrf54lm20a/cpuapp
+    - nrf54l20pdk/nrf54l20/cpuflpr
```
`nrf54l20pdk`, both cpuapp and cpuflpr were deleted in mean time. Please remove.

nordic-krch commented:
@anangl , @masz-nordic can you take a look?

anangl commented on drivers/timer/nrf_grtc_timer.c:
```diff
@@ -555,18 +557,48 @@ void sys_clock_set_timeout(int32_t ticks, bool idle)
 
-	/* Rounded down target_time to the tick boundary
-	 * (but not less than one tick after the last)
+	cc_value = last_count + last_elapsed + cyc;
```
I think the following will be clearer, and `now` is anyway needed a few lines below:
```suggestion
	int64_t now = last_count + last_elapsed;

	cc_value = now + cyc;
```

anangl commented on drivers/timer/nrf_grtc_timer.c:
```diff
@@ -456,7 +452,9 @@ uint32_t sys_clock_elapsed(void)
 	}
 
-	return (uint32_t)(counter_sub(counter(), last_count) / CYC_PER_TICK);
+	last_elapsed = (uint32_t)counter_sub(counter(), last_count);
```
Is it safe to assume that elapsed ticks will always fit into a 32-bit variable (for 1 MHz clock, it's maximum 4294 seconds)?

anangl commented:
I think the PR needs to be rebased to solve the twister failure:
```
Run export ZEPHYR_BASE=${PWD}
Traceback (most recent call last):
  File "/home/runner/work/zephyr/zephyr/zephyr/./scripts/ci/test_plan.py", line 38, in <module>
    import list_boards
  File "/home/runner/work/zephyr/zephyr/zephyr/scripts/list_boards.py", line 13, in <module>
    import jsonschema
ModuleNotFoundError: No module named 'jsonschema'
```

anangl commented on drivers/timer/nrf_grtc_timer.c:
Why do you need this casting?

nordic-krch replied:
it might be negative

nordic-krch replied:
in case of abort

ppelikan-nordic committed:
tests: kernel: timer: behavior: Adjustments for NRF54H20 PPR

Adjusting CONFIG_SYS_CLOCK_TICKS_PER_SEC as core is to slow too run
the test with the default system clock frequency.

Signed-off-by: Paweł Pelikan <pawel.pelikan@nordicsemi.no>
